### PR TITLE
[CALCITE-4526] Fixup the unparse's output of SqlSnapshot,when tableRel is a SqlAsOperator (jibiyr)

### PR DESCRIPTION
**Pull request by jibiyr**
before this change when the sqlSnapshot' tableRel is a alias. the unparse result is wrong.
add test for sqlSnapshot and fix it.

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/sql/parser/SqlParserTest.java [hunk 1]
```diff
@@ -7766,6 +7770,32 @@ public void subTestIntervalSecondFailsValidation() {
...
+        new SqlParserPos(8, 11)
+    );
+    SqlNode period = new SqlIdentifier("orders.proctime", new SqlParserPos(8, 42));
+    SqlSnapshot sqlSnapshot = new SqlSnapshot(new SqlParserPos(8, 11), tableRef, period);
+
+    SqlWriter sqlWriter = new SqlPrettyWriter();
+    sqlSnapshot.unparse(sqlWriter, 0, 0);
+
+    assertEquals(sqlWriter.toSqlString().getSql(), "\"default_db.products\" "
+        + "FOR SYSTEM_TIME AS OF \"orders.proctime\" AS \"products1\"");
```
[the remark this answers is not in the document]
Comment (jibiyr) [reply core/src/test/java/org/apache/calcite/sql/parser/SqlParserTest.java:7789]: thank you, danny I'll change it

**Comment by jibiyr** [comment]
I have change ,please review again @danny0405

File: core/src/main/java/org/apache/calcite/sql/SqlSnapshot.java [hunk 2]
```diff
@@ -127,10 +129,25 @@ private SqlSnapshotOperator() {
...
+        writer.keyword("AS");
+        basicCall.operand(1).unparse(writer, 0, 0);
+      } else {
+        tableRef.unparse(writer, 0, 0);
+        writeKeywordAndPeriod(writer, snapshot);
+      }
     }
   }
+
+  private static void writeKeywordAndPeriod(SqlWriter writer, SqlSnapshot snapshot) {
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/sql/SqlSnapshot.java:149]:
```suggestion
  private static void writeForSustemTimeAsOf(SqlWriter writer, SqlSnapshot snapshot) {
```

**Review by danny0405 — APPROVED** [review]
+1, looks good from my side.

File: core/src/test/java/org/apache/calcite/sql/parser/SqlParserTest.java [hunk 3]
```diff
@@ -7766,6 +7766,15 @@ public void subTestIntervalSecondFailsValidation() {
         SqlParserUtil.addCarets("abcdef", 1, 7, 1, 7));
   }
 
+  @Test void testSnapshot() {
```
Comment (vlsi) [inline core/src/test/java/org/apache/calcite/sql/parser/SqlParserTest.java:7769]:
```suggestion
  @Test void testSnapshotForSystemTimeWithAlias() {
```

Comment (jibiyr) [reply core/src/test/java/org/apache/calcite/sql/parser/SqlParserTest.java]: I have changed

File: core/src/main/java/org/apache/calcite/sql/SqlSnapshot.java [hunk 4]
```diff
@@ -127,10 +129,25 @@ private SqlSnapshotOperator() {
...
+      } else {
+        tableRef.unparse(writer, 0, 0);
+        writeForSystemTimeAsOf(writer, snapshot);
+      }
     }
   }
+
+  private static void writeForSystemTimeAsOf(SqlWriter writer, SqlSnapshot snapshot) {
+    writer.keyword("FOR SYSTEM_TIME AS OF");
+    snapshot.period.unparse(writer, 0, 0);
```
Comment (chunweilei) [inline core/src/main/java/org/apache/calcite/sql/SqlSnapshot.java:151]: Maybae it is better to put `writeForSystemTimeAsOf` in SqlSnapshotOperator.

Comment (jibiyr) [reply core/src/main/java/org/apache/calcite/sql/SqlSnapshot.java:151]: ok,I'll change it

**Comment by jibiyr** [comment]
@chunweilei  anymore suggestion, and what should I do next?

**Pull request closed without merging**